app/gui2/src/components/GraphEdge.vue

```diff
@@ -1,57 +1,501 @@
 <script setup lang="ts">
 import type { Edge } from '@/stores/graph'
 import type { Rect } from '@/util/rect'
+import { Vec2 } from '@/util/vec2'
 import { clamp } from '@vueuse/core'
 import type { ExprId } from 'shared/yjsModel'
-import { computed } from 'vue'
+import { computed, ref } from 'vue'
 
 const props = defineProps<{
   edge: Edge
   nodeRects: Map<ExprId, Rect>
   exprRects: Map<ExprId, Rect>
   exprNodes: Map<ExprId, ExprId>
+  hoveredNode: ExprId | undefined
+  hoveredExpr: ExprId | undefined
+  sceneMousePos: Vec2 | null
 }>()
 
-const edgePath = computed(() => {
-  let edge = props.edge
-  const targetNodeId = props.exprNodes.get(edge.target)
-  if (targetNodeId == null) return ''
-  let sourceNodeRect = props.nodeRects.get(edge.source)
-  let targetNodeRect = props.nodeRects.get(targetNodeId)
-  let targetRect = props.exprRects.get(edge.target)
-  if (sourceNodeRect == null || targetRect == null || targetNodeRect == null) return ''
-  let sourcePos = sourceNodeRect.center()
-  let targetPos = targetRect.center().add(targetNodeRect.pos)
-
-  let sourceRangeX = sourceNodeRect.rangeX()
-  const EDGE_PADDING = 20
-  const sourceX = clamp(targetPos.x, sourceRangeX[0] + EDGE_PADDING, sourceRangeX[1] - EDGE_PADDING)
-  const LINE_OUT = 20
-  const QUAD_OUT = 50
-
-  const midpointX = (sourceX + targetPos.x) / 2
-  const midpointY = (sourcePos.y + targetPos.y) / 2
-
-  return `
-    M ${sourceX} ${sourcePos.y}
-    L ${sourceX} ${sourcePos.y + LINE_OUT}
-    Q ${sourceX} ${sourcePos.y + QUAD_OUT} ${midpointX} ${midpointY}
-    Q ${targetPos.x} ${targetPos.y - QUAD_OUT} ${targetPos.x} ${targetPos.y - LINE_OUT}
-    L ${targetPos.x} ${targetPos.y}
-  `
+const emit = defineEmits<{
+  disconnectSource: []
+  disconnectTarget: []
+}>()
+
+const base = ref<SVGPathElement>()
+
+function targetPos(): { pos: Vec2; size?: Vec2 } | null {
+  const hovered = props.hoveredNode != props.edge.source ? props.hoveredExpr : undefined
+  const targetExpr = props.edge.target ?? hovered
+  if (targetExpr != null) {
+    const targetNodeId = props.exprNodes.get(targetExpr)
+    if (targetNodeId == null) return null
+    const targetNodeRect = props.nodeRects.get(targetNodeId)
+    const targetRect = props.exprRects.get(targetExpr)
+    if (targetRect == null || targetNodeRect == null) return null
+    return { pos: targetRect.center().add(targetNodeRect.pos), size: targetRect.size }
+  } else if (props.sceneMousePos != null) {
+    return { pos: props.sceneMousePos }
+  } else {
+    return null
+  }
+}
+function sourcePos(): { pos: Vec2; size?: Vec2 } | null {
+  const targetNode = props.edge.target != null ? props.exprNodes.get(props.edge.target) : undefined
+  const hovered = props.hoveredNode != targetNode ? props.hoveredNode : undefined
+  const sourceNode = props.edge.source ?? hovered
+  if (sourceNode != null) {
```

Comment:
In some situations many reactive values in those expressions are accessed only to get thrown away later. That still registers as data dependency. That makes `sourcePos` of **all** edges depend on `hoveredNode` prop, even when they have both edge target and source set. To prevent that, the `props.edge.source` null check should be done before anything else. Same things applies to `targetPos` function above.